### PR TITLE
Windows: implement readSymbolicLink

### PR DESCRIPTION
**Pull request by laszlocsomor**
In this PR:

- Implement ReadSymlinkOrJunction in the JNI
  library, and expose it to Java code. This method
  can resolve symlinks and junctions.

- Override readSymbolicLink in WindowsFileSystem
  and use the new native method

- Add unit tests to exercise junction resolution.
  Manual testing shows the code works for symlinks
  too, however we cannot reliably create them in
  tests because that requires unprivileged symlink
  creation support (i.e. for the test to run as
  Admin, or for the OS having enabled it)

Fixes https://github.com/bazelbuild/bazel/issues/7907

### CODE REVIEW

**Comment by laszlocsomor** [comment]
This PR is ready, but depends on https://github.com/bazelbuild/bazel/pull/7928.

File: src/main/java/com/google/devtools/build/lib/windows/jni/WindowsFileOperations.java [hunk 1]
```diff
@@ -65,12 +65,22 @@ private WindowsFileOperations() {
   private static final int DELETE_PATH_DIRECTORY_NOT_EMPTY = 3;
   private static final int DELETE_PATH_ACCESS_DENIED = 4;
 
+  private static final int READ_SYMLINK_OR_JUNCTION_SUCCESS = 0;
+  // READ_SYMLINK_OR_JUNCTION_ERROR = 1;
```
Comment (meteorcloudy) [inline src/main/java/com/google/devtools/build/lib/windows/jni/WindowsFileOperations.java]: Why comment here? Is `READ_SYMLINK_OR_JUNCTION_ERROR` set or used anywhere?

Comment (laszlocsomor) [reply src/main/java/com/google/devtools/build/lib/windows/jni/WindowsFileOperations.java]: For documentation, and to indicate that the value exists. The variable is not used anywhere (handled by `switch.default`).

Removed.

**Comment by laszlocsomor** [comment]
Thanks for the pointer. Yes, I'll refactor `src/main/cpp/util/file_windows.cc`.

I'd rather not expect symlink support in tests. I tend to run tests locally, and I have symlinks disabled (to emulate users' environment).

File: src/main/java/com/google/devtools/build/lib/windows/jni/WindowsFileOperations.java [hunk 2]
```diff
@@ -66,7 +66,6 @@ private WindowsFileOperations() {
   private static final int DELETE_PATH_ACCESS_DENIED = 4;
 
   private static final int READ_SYMLINK_OR_JUNCTION_SUCCESS = 0;
-  // READ_SYMLINK_OR_JUNCTION_ERROR = 1;
   private static final int READ_SYMLINK_OR_JUNCTION_ACCESS_DENIED = 2;
```
Comment (meteorcloudy) [inline src/main/java/com/google/devtools/build/lib/windows/jni/WindowsFileOperations.java]: Nit: maybe 1, 2, 3, 4,?

Comment (laszlocsomor) [reply src/main/java/com/google/devtools/build/lib/windows/jni/WindowsFileOperations.java]: These values match constants in `src/main/native/windows/file.h`. Added comments.

**Pull request closed without merging**